[OR-308] Add instructions for using PostgreSQL (#10)
PegaSysEng/doc.orion: docs/Configuring-Orion/Using-PostgreSQL.md

description: Storing Information with PostgreSQL 
<!--- END of page meta data -->

# Store Information with PostgreSQL

> PostgreSQL is an object-relational database management system you can use to store key values for Orion.

Reviewer: I'd update _key values_ to _private transaction payloads_ so it's more specific

Reviewer: The point of using a relational database rather than the default is for disaster recovery so I'd add a link to that section.

Author: Fixed

> PostgreSQL is an object-relational database management system you can use to store key values for Orion.

Reviewer 2: Not sure what is meant by management system, so would drop "management system" from the description.

Author: I found that phrase in the description of PostgreSQL. Happy to remove it.

Author: Removed

## Prerequisites

* [PostgreSQL](https://www.postgresql.org/)

## Set up your PostgreSQL database

1. In your database, create the `store` table by using the provided DDL script, as shown in the following example: 

>     `psql -h HOST-NAME -U USER-NAME -d DATABASE-NAME -f ./docs/Configuring-Orion/database/postgres_storage.sql` 

Reviewer: I'm not sure if we have this doc'd in our conventions but we usually use angle brackets around things that need replacing. eg: 
`psql -h <HOST-NAME> -U <USER-NAME> -d <DATABASE-NAME>`

And Bryon and I were talking about this the other day and thought it's probably worth calling it out explicitly when things need replacing to prevent users copying and pasting without replacing them. eg under the example: 
Where `<HOST-NAME>`, `<USER-NAME>`, and `<DATABASE-NAME>` are replaced by the appropriate values for your database.

Author: Fixed

>     `psql -h HOST-NAME -U USER-NAME -d DATABASE-NAME -f ./docs/Configuring-Orion/database/postgres_storage.sql` 

Reviewer 2: Should the ddl file also be in this docs repo? If not we should point out that user needs to access the ddl file from the main Orion repo.

Author: Changed to point to the location in the main Orion repo

> 2. In your database, specify the users for the `store` table and assign permissions to them.

Reviewer: _In your database_ is repeated at the start of both steps.  Is it worth pulling it out to remove the repetition? eg: 

In your database: 

1.  Create the ....
2. Specify the ...

Author: Changed.

## Configure Orion to use PostgreSQL

> In your [Orion configuration file](Configuration-File.md), specify `postgresql` for the `storage` property, as shown in the following example. Substitute your own values for `DATABASE-NAME`, `USER-NAME`, and `PASSWORD`. 

Reviewer: After putting my long comment above about calling out things that need replacing, you've done that here :-)

> In your [Orion configuration file](Configuration-File.md), specify `postgresql` for the `storage` property, as shown in the following example. Substitute your own values for `DATABASE-NAME`, `USER-NAME`, and `PASSWORD`. 

Reviewer 2: Should USER-NAME be hyphenated?

  ```

>   storage="sql:jdbc:postgresql://localhost:5432/<DATABASE-NAME>?user=<USER-NAME>&password=<PASSWORD>"

Reviewer: Does localhost:5432 also need replacing here?  Is that value specific to your setup?

Author: Changed these to variables

>   storage="sql:jdbc:postgresql://localhost:5432/<DATABASE-NAME>?user=<USER-NAME>&password=<PASSWORD>"

Reviewer 2: The host i.e. localhost and the port 5432 should also be split out as values to be substituted.

Author: Fixed

  ```
  

> ## Test your updated configuration

Reviewer: The numbering in this section isn't displaying correctly - it'll be because the code blocks need indenting I imagine

Author: Fixed

description: Storing Information with PostgreSQL 
<!--- END of page meta data -->

# Store Information with PostgreSQL

> PostgreSQL is an object-relational database you can use to store private transaction payloads for Orion. Consider using this option as an aid in [disaster recovery](Disaster-Recovery.md).

Reviewer: ```suggestion
PostgreSQL is an object-relational database you can use to store private transaction payloads for Orion. Use a relational database to support your [disaster recovery strategy](Disaster-Recovery.md).
```

Suggested update to make more active.

## Prerequisites

* [PostgreSQL](https://www.postgresql.org/)

## Set up your PostgreSQL database

In your database:

> 1. Create the `store` table by using the provided DDL script, as shown in the following example. Substute your own values for `HOST-NAME`, `USER-NAME`, and `DATABASE-NAME`.

Reviewer: Even thought the script is now provided with the binary I think we're going to need to tell them where it is.  @jframe -> will in be in the location where the binary is installed?

Reviewer 3: @MadelineMurray <installation_dir>/orion/database/postgres_ddl.sql

>     `psql -h <HOST-NAME> -U <USER-NAME> -d <DATABASE-NAME> -f ~/orion/docs/configuring-orion/database/postgres_storage.sql` 

Reviewer 2: For this to work, the user will have to have the Orion source code. So think we should add a step to download Orion source code or clone the Orion repo before running psql.
A thought, perhaps the ddl scripts should be included in the Orion distribution?

Reviewer 4: Forcing to download all Orion source code just ot have the DDL scripts seems a bit overkill...

Author: The DDL file will be in the Orion distribution. See https://pegasys1.atlassian.net/browse/OR-326.

>     `psql -h <HOST-NAME> -U <USER-NAME> -d <DATABASE-NAME> -f ~/orion/docs/configuring-orion/database/postgres_storage.sql` 

Reviewer: The path to `progres_storage.sql` will need to be updated to show it'll have to be replaced to the path the script (ie, the path you've given assumes that you're using the source code doesn't it?)

    
2. Specify the users for the `store` table and assign permissions to them.

## Configure Orion to use PostgreSQL

In your [Orion configuration file](Configuration-File.md), specify `postgresql` for the `storage` property, as shown in the following example. Substitute your own values for `HOST`, `PORT`, `DATABASE-NAME`, `USER-NAME`, and `PASSWORD`. 

  ```
  storage="sql:jdbc:postgresql://<HOST>:<PORT>/<DATABASE-NAME>?user=<USER-NAME>&password=<PASSWORD>"
  ```
  
## Test your updated configuration

1. Send the payload to Orion by running the following request:

     ```
     curl -X POST \
     http://localhost:8888/send \
     -H 'Content-Type: application/json' \
     -d '{ 
           "payload": "SGVsbG8sIFdvcmxkIQ==",

>            "from": "<KEY-IN-NODEKEY>.PUB",

Reviewer 3: there is no mention of .PUB in https://docs.orion.pegasys.tech/en/stable/Getting-Started/Quickstart/ ... shouldn't we use similar content as in quick start here?

Author: Let me make sure I understand your comment. Are you suggesting that the doc should say `<OrionPublicKey>` instead of `<KEY-IN-NODEKEY.PUB>`?

           "to": ["<KEY-IN-NODEKEY>.PUB"]
          }'
      ```

1. Using psql, the terminal-based front-end to PostgreSQL, retrieve the new row in the `store` table.

1. Receive the payload by running the following request:

     ```
      curl -X POST \
      http://localhost:8888/receive \
      -H 'Content-Type: application/json' \
      -d '{
           "key": "<KEY-RETURNED-IN-STEP-1>",
            "to": "<KEY-IN-NODEKEY>.PUB"
          }'
      ```
  

> 1. Using psql, retrieve the data in the `store` table.

Reviewer 3: what is the difference between step 2 and 4? both would return same data from store database.

Author: Step 2 should return the empty row in the store table. Step 4 should return the data in the row. Now that I'm typing the explanation, Step 2 sounds unnecessary.

Reviewer 3: when I ran step 2, I already had two entries in store ... before using any curl command.
